From f5abe21b90490588a52888005f8fb9f044ea02c7 Mon Sep 17 00:00:00 2001
From: Wouter van Kesteren <woutershep@gmail.com>
Date: Sat, 16 Apr 2016 14:37:51 +0200
Subject: [PATCH] Ban unprefixed thingies infrastructure.

Adds four new things:

A BANNEDDIR environment variable with location determined by config.
This directory will be prepended to the PATH if set.

A banned_by_distribution util, to make the little stubs survive a
freezing of exheres-0 to exheres-1.

A dobanned that creates little stubs in ${IMAGE}/${BANNEDDIR} that
call banned_by_distribution properly. The stubs are then potentially
shuffled around by the alternatives infrastructure.

BANNEDDIR must not be changed and must not be empty

Change-Id: Ie38703f8922c19a30924a928ddf7e20b24ffa5d0
---

diff --git a/paludis/repositories/e/e_repository_TEST_exheres_0.cc b/paludis/repositories/e/e_repository_TEST_exheres_0.cc
index 1104484..48ad545 100644
--- a/paludis/repositories/e/e_repository_TEST_exheres_0.cc
+++ b/paludis/repositories/e/e_repository_TEST_exheres_0.cc
@@ -571,6 +571,22 @@
         ASSERT_TRUE(bool(id));
         EXPECT_THROW(id->perform_action(action), ActionFailedError);
     }
+
+    {
+        const std::shared_ptr<const PackageID> id(*env[selection::RequireExactlyOne(generator::Matches(
+                        PackageDepSpec(parse_user_package_dep_spec("=cat/banned-0",
+                                &env, { })), nullptr, { }))]->last());
+        ASSERT_TRUE(bool(id));
+        EXPECT_THROW(id->perform_action(action), ActionFailedError);
+    }
+
+    {
+        const std::shared_ptr<const PackageID> id(*env[selection::RequireExactlyOne(generator::Matches(
+                        PackageDepSpec(parse_user_package_dep_spec("=cat/banned-1",
+                                &env, { })), nullptr, { }))]->last());
+        ASSERT_TRUE(bool(id));
+        id->perform_action(action);
+    }
 }
 
 TEST(ERepository, ReallyInstallExheres0)
diff --git a/paludis/repositories/e/e_repository_TEST_exheres_0_setup.sh b/paludis/repositories/e/e_repository_TEST_exheres_0_setup.sh
index a9bb846..61cb4b4 100755
--- a/paludis/repositories/e/e_repository_TEST_exheres_0_setup.sh
+++ b/paludis/repositories/e/e_repository_TEST_exheres_0_setup.sh
@@ -1098,6 +1098,40 @@
     expecting_tests --expensive
 }
 END
+mkdir -p "packages/cat/banned"
+cat <<'END' > packages/cat/banned/banned-0.ebuild || exit 1
+DESCRIPTION="The Long Description"
+SUMMARY="The Short Description"
+HOMEPAGE="http://example.com/"
+DOWNLOADS=""
+SLOT="0"
+MYOPTIONS="spork"
+LICENCES="GPL-2"
+PLATFORMS="test"
+WORK="${WORKBASE}"
+
+pkg_setup() {
+    edo banned_by_distribution illegal-executable
+}
+END
+cat <<'END' > packages/cat/banned/banned-1.ebuild || exit 1
+DESCRIPTION="The Long Description"
+SUMMARY="The Short Description"
+HOMEPAGE="http://example.com/"
+DOWNLOADS=""
+SLOT="0"
+MYOPTIONS="spork"
+LICENCES="GPL-2"
+PLATFORMS="test"
+WORK="${WORKBASE}"
+
+src_install() {
+    [[ -n ${BANNEDDIR} ]] || die
+
+    dobanned illegal-executable
+    [[ -x ${IMAGE}/${BANNEDDIR}/illegal-executable ]] || die
+}
+END
 mkdir -p "packages/cat/exdirectory-phase"
 cat <<'END' > packages/cat/exdirectory-phase/exdirectory-phase-1.ebuild || exit 1
 DESCRIPTION="The Long Description"
diff --git a/paludis/repositories/e/eapi.cc b/paludis/repositories/e/eapi.cc
index d1ab40e..f3e2e00 100644
--- a/paludis/repositories/e/eapi.cc
+++ b/paludis/repositories/e/eapi.cc
@@ -167,6 +167,7 @@
     std::shared_ptr<const EAPIEbuildOptions> make_ebuild_options(const KeyValueConfigFile & k)
     {
         return std::make_shared<EAPIEbuildOptions>(make_named_values<EAPIEbuildOptions>(
+                        n::banneddir() = check_get(k, "banneddir"),
                         n::bash_compat() = check_get(k, "bash_compat"),
                         n::binary_from_env_variables() = check_get(k, "binary_from_env_variables"),
                         n::bracket_merged_variables() = check_get(k, "bracket_merged_variables"),
diff --git a/paludis/repositories/e/eapi.hh b/paludis/repositories/e/eapi.hh
index 9954512..ea4cc45 100644
--- a/paludis/repositories/e/eapi.hh
+++ b/paludis/repositories/e/eapi.hh
@@ -40,6 +40,7 @@
     {
         typedef Name<struct name_allow_tokens_in_mask_files> allow_tokens_in_mask_files;
         typedef Name<struct name_annotations> annotations;
+        typedef Name<struct name_banneddir> banneddir;
         typedef Name<struct name_bash_compat> bash_compat;
         typedef Name<struct name_best_has_version_host_root> best_has_version_host_root;
         typedef Name<struct name_binary_from_env_variables> binary_from_env_variables;
@@ -420,6 +421,7 @@
 
         struct EAPIEbuildOptions
         {
+            NamedValue<n::banneddir, std::string> banneddir;
             NamedValue<n::bash_compat, std::string> bash_compat;
             NamedValue<n::binary_from_env_variables, std::string> binary_from_env_variables;
             NamedValue<n::bracket_merged_variables, std::string> bracket_merged_variables;
diff --git a/paludis/repositories/e/eapis/exheres-0.conf b/paludis/repositories/e/eapis/exheres-0.conf
index a5a540a..3b08e03 100644
--- a/paludis/repositories/e/eapis/exheres-0.conf
+++ b/paludis/repositories/e/eapis/exheres-0.conf
@@ -45,6 +45,8 @@
     +/etc -/etc/binfmt.d -/etc/modules-load.d -/etc/sysctl.d -/etc/tmpfiles.d \
           -/etc/systemd/system -/etc/udev/rules.d -/etc/bash_complection.d
 
+banneddir = /usr/share/exherbo/banned_by_distribution
+
 vdb_from_env_variables = \
     CATEGORY CHOST DEPENDENCIES SUMMARY EAPI \
     HOMEPAGE INHERITED MYOPTIONS PLATFORMS LICENCES PNVR \
@@ -70,7 +72,7 @@
 non_empty_variables = \
     PNV PV PR PN PVR PNVR CATEGORY build:FILES build:EXLIBSDIRS build:FETCHEDDIR build:REPODIR \
     PALUDIS_TMPDIR PALUDIS_EBUILD_LOG_LEVEL PALUDIS_EBUILD_DIR \
-    ROOT PALUDIS_PACKAGE_BUILDDIR
+    ROOT PALUDIS_PACKAGE_BUILDDIR BANNEDDIR
 
 directory_variables = \
     build:FETCHEDDIR build:REPODIR
@@ -131,7 +133,8 @@
     PALUDIS_DIRECTORY_IF_EXISTS_VARIABLES PALUDIS_SOURCE_MERGED_VARIABLES \
     PALUDIS_MUST_NOT_CHANGE_VARIABLES PALUDIS_RDEPEND_DEFAULTS_TO_DEPEND \
     ECLASSDIR ECLASSDIRS EXLIBSDIRS PALUDIS_VARIABLE PALUDIS_PROFILE_DIR \
-    PALUDIS_PROFILE_DIRS PALUDIS_LOADSAVEENV_DIR PALUDIS_BRACKET_MERGED_VARIABLES
+    PALUDIS_PROFILE_DIRS PALUDIS_LOADSAVEENV_DIR \
+    PALUDIS_BRACKET_MERGED_VARIABLES BANNEDDIR
 
 load_modules = \
     conditional_functions kernel_functions sydbox portage_stubs ever_functions \
diff --git a/paludis/repositories/e/ebuild.cc b/paludis/repositories/e/ebuild.cc
index f0a13fc..b2cf19a 100644
--- a/paludis/repositories/e/ebuild.cc
+++ b/paludis/repositories/e/ebuild.cc
@@ -335,6 +335,9 @@
         process.setenv("PALUDIS_CROSS_COMPILE_TOOL_PREFIX",
                        params.tool_prefix());
 
+    if (! options->banneddir().empty())
+        process.setenv("BANNEDDIR", options->banneddir());
+
     if (options->want_portage_emulation_vars())
         add_portage_vars(process);
 
diff --git a/paludis/repositories/e/ebuild/ebuild.bash b/paludis/repositories/e/ebuild/ebuild.bash
index 8a2d4d8..82370a0 100755
--- a/paludis/repositories/e/ebuild/ebuild.bash
+++ b/paludis/repositories/e/ebuild/ebuild.bash
@@ -54,6 +54,8 @@
     # See ticket:374.
     export PATH="/usr/bin:/usr/sbin:/bin:/sbin${PATH:+:${PATH}}"
 
+    [[ -n "${BANNEDDIR}" ]] && export PATH="${BANNEDDIR}:${PATH}"
+
     export PATH="${PALUDIS_EBUILD_DIR}/utils:${PATH}"
     # Automake likes to scatter our utilities over two directories.
     [[ -n "${PALUDIS_EBUILD_DIR_FALLBACK}" ]] && export PATH="${PALUDIS_EBUILD_DIR_FALLBACK}/utils:${PATH}"
diff --git a/paludis/repositories/e/ebuild/utils/exheres-0/Makefile.am b/paludis/repositories/e/ebuild/utils/exheres-0/Makefile.am
index 6b3156f..1ac60e6 100644
--- a/paludis/repositories/e/ebuild/utils/exheres-0/Makefile.am
+++ b/paludis/repositories/e/ebuild/utils/exheres-0/Makefile.am
@@ -34,7 +34,9 @@
 	ecompress \
 	ecompressdir \
 	strip \
-	banned_in_eapi_exheres-0
+	banned_in_eapi_exheres-0 \
+	dobanned \
+	banned_by_distribution
 
 AM_CXXFLAGS = -I$(top_srcdir) @PALUDIS_CXXFLAGS@
 
diff --git a/paludis/repositories/e/ebuild/utils/exheres-0/banned_by_distribution b/paludis/repositories/e/ebuild/utils/exheres-0/banned_by_distribution
new file mode 100755
index 0000000..45c9a53
--- /dev/null
+++ b/paludis/repositories/e/ebuild/utils/exheres-0/banned_by_distribution
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
+# vim: set sw=4 sts=4 et :
+
+# Copyright (c) 2007 Ciaran McCreesh
+# Copyright (c) 2016 Wouter van Kesteren
+#
+# This file is part of the Paludis package manager. Paludis is free software;
+# you can redistribute it and/or modify it under the terms of the GNU General
+# Public License as published by the Free Software Foundation; either version
+# 2 of the License, or (at your option) any later version.
+#
+# Paludis is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
+# Place, Suite 330, Boston, MA  02111-1307  USA
+
+COLOUR_RED=$'\e[31;01m'
+COLOUR_NORMAL=$'\e[0m'
+
+echo "${COLOUR_RED}!!! Exheres bug: '${1}' banned by distribution${COLOUR_NORMAL}"
+
+exit 123
+
diff --git a/paludis/repositories/e/ebuild/utils/exheres-0/dobanned b/paludis/repositories/e/ebuild/utils/exheres-0/dobanned
new file mode 100755
index 0000000..f9dfee9
--- /dev/null
+++ b/paludis/repositories/e/ebuild/utils/exheres-0/dobanned
@@ -0,0 +1,61 @@
+#!/usr/bin/env bash
+# vim: set sw=4 sts=4 et :
+
+# Copyright (c) 2006 Stephen Bennett
+# Copyright (c) 2016 Wouter van Kesteren
+#
+# Based in part upon dosbin from Portage, which is Copyright 1995-2005
+# Gentoo Foundation and distributed under the terms of the GNU General
+# Public License v2.
+#
+# This file is part of the Paludis package manager. Paludis is free software;
+# you can redistribute it and/or modify it under the terms of the GNU General
+# Public License as published by the Free Software Foundation; either version
+# 2 of the License, or (at your option) any later version.
+#
+# Paludis is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
+# Place, Suite 330, Boston, MA  02111-1307  USA
+
+source "${PALUDIS_EBUILD_DIR}"/die_functions.bash
+
+if [[ ! -d ${!PALUDIS_IMAGE_DIR_VAR} ]]; then
+    paludis_die_or_error "\${${PALUDIS_IMAGE_DIR_VAR}} not valid; aborting"
+fi
+
+if [[ -z ${BANNEDDIR} ]]; then
+    paludis_die_or_error "\${BANNEDDIR} not valid; aborting"
+fi
+
+if [[ ${#} -lt 1 ]]; then
+    paludis_die_or_error "at least one argument needed"
+fi
+
+if [[ ! -d "${!PALUDIS_IMAGE_DIR_VAR}/${BANNEDDIR}" ]]; then
+    install -d "${!PALUDIS_IMAGE_DIR_VAR}/${BANNEDDIR}" || \
+        paludis_die_or_error "could not create ${!PALUDIS_IMAGE_DIR_VAR}/${BANNEDDIR}"
+fi
+
+ret=0
+for x in "${@}"; do
+    rm -rf "${!PALUDIS_TEMP_DIR_VAR}/${x}"
+
+    cat > "${!PALUDIS_TEMP_DIR_VAR}/${x}" <<'EOF' || ret=1
+#!/bin/sh
+exec banned_by_distribution "${0##*/}"
+EOF
+
+    if [[ -n ${PALUDIS_NO_CHOWN} ]]; then
+        install -m0755 "${!PALUDIS_TEMP_DIR_VAR}/${x}" "${!PALUDIS_IMAGE_DIR_VAR}/${BANNEDDIR}" || ret=2
+    else
+        install -m0755 -o root -g 0 "${!PALUDIS_TEMP_DIR_VAR}/${x}" "${!PALUDIS_IMAGE_DIR_VAR}/${BANNEDDIR}" || ret=2
+    fi
+done
+
+[[ 0 != "${ret}" ]] && paludis_die_or_error "dobanned returned error ${ret}"
+exit ${ret}
